File: scala/gildedrose/src/test/scala/InventorySpec.scala
import org.scalatest._

class InventorySpec extends FunSpec with Matchers {

  describe("updateQuality") {
    describe("basic item") {
      describe("when quality and sell_in is positive") {
        it("reduces sell_in and quality by 1") {
          val item = Item("Foo", sell_in = 10, quality = 10)

          var inventory = Inventory(Seq(item))
          inventory = inventory.update_quality

          assert(inventory.items.head.sell_in == 9)
          assert(inventory.items.head.quality == 9)
        }
      }

      describe("when sell by date has passed") {
        it("reduces quality by half") {
          val item = Item("Foo", sell_in = 0, quality = 10)

          var inventory = Inventory(Seq(item))
          inventory = inventory.update_quality

          assert(inventory.items.head.sell_in == 0)
          assert(inventory.items.head.quality == 5)
        }
      }

      describe("when a quality is already 0") {
        it("maintains the quality as 0") {
          val item = Item("Foo", sell_in = 10, quality = 0)

          var inventory = Inventory(Seq(item))
          inventory = inventory.update_quality

          assert(inventory.items.head.sell_in == 9)
          assert(inventory.items.head.quality == 0)
        }
      }
    }

    describe("Increase by age item") {

    }
  }
}

File: ruby/gildedrose/test/inventory_spec.rb
require 'inventory'

RSpec.describe Inventory do
  describe "add_item" do
    it "adds a item to the list" do
      inventory = Inventory.new
      new_item = Item.new("Foo", 1, 50, BasicItemType.new)
      inventory.add_item(new_item)
      expect(inventory.items.size).to eq(1)
    end
  end

  describe "update_quality" do
    it "updates the quality of all items" do
      inventory = Inventory.new

      item1 = Item.new("Foo", 1, 50, BasicItemType.new)
      item2 = Item.new("Foo", 10, 10, BasicItemType.new)

      inventory.add_item(item1)
      inventory.add_item(item2)

      expect(item1).to receive(:update_quality)
      expect(item2).to receive(:update_quality)

      inventory.update_quality
    end
  end
end
